### ISSUE TITLE
CanvasControl memory leak?

### ISSUE THREAD

**Issue by JohnnyWestlake** [Microsoft/Win2D#37]
I've been trying to track down a memory leak in my application, and I seem to have tracked it down to something to do with the CanvasControl

I've reproduced the issue somewhat in a *very* basic example application here. If you enabled Profile in VS and switch backwards and forwards between the two pages, you'll see application usage grow (albeit slightly) each time. This get exacerbated in my actual application when I'm rendering actual content in the CanvasControl and on the surrounding page.

(I can't really see anything helpful in the memory profiler either, other than "Overlapped Data" - I have no diea what that is. You can also observe a memory leak in the sample app - navigate forward and backwards between the hubpage and the shape / stroke styles pages and watch the memory grow, although I've not looked at whether there are any obvious causes for that particular leak)

Very simple sample project here:

https://onedrive.live.com/redir?resid=DE47921B27CEAD6E!149144&authkey=!AKVhsU_ebScBXuw&ithint=file%2czip

Is this reproducible on your end?
Thanks!

**Comment by damyanp** [comment]
Thanks for reporting this with such a great minimal repro - I can confirm that we also see the leak.  You can see slightly more information if you run the memory profiler in mixed native/managed mode and even more information if you have symbols for Microsoft.Graphics.Canvas.dll (eg if you build it from source).

What I believe is going on here is that the CanvasControl isn't getting destroyed / garbage collected.  Instead, XAML sends it an Unloaded event which we should be responding to by freeing up the native resources that CanvasControl has allocated.  We haven't implemented that yet.

We'll work on resolving this soon.

Thank you again!

**Comment by shawnhar** [comment]
Hi,

We made some fixes in the last sprint (code just released today) which partially improve the situation.  There was a reference count cycle, now fixed, that meant CanvasControl was never getting reclaimed even after you released all application side references to it  (oops :-)

Another piece of work we are planning for this sprint will make the situation even better:  we intend to destroy the graphics device when the control is unloaded from the XAML visual graph, even if the control itself is still referenced.  We'll create a new device if the control later gets re-added, but in most cases we expect this to result in memory being reclaimed sooner as we'll no longer be dependent on nondeterministic garbage collection to dictate when expensive resources such as the underlying D3D device are destroyed.

In short: this issue should be improved in the current version, but we're keeping it open until the last part of the fix is in.

**Comment by shawnhar** [comment]
Win2D version 0.0.13 (released today) includes some more fixes that address all but one of our memory leak problems.  The remaining issue is turning out to be more difficult than expected, but we're still working on it.

In the meantime, a workaround is to manually unhook all event handlers from the CanvasControl or CanvasAnimatedControl before navigating away from the page.  The problem is that these event handlers cause a reference count cycle, and manually removing the events will break that cycle so the control can be properly garbage collected.

For instance you could use the Unloaded event of your XAML page to unhook CreateResources, Draw, Update, SizeChanged, any pointer input events, etc, that you may have previously subscribed to.

Stay tuned for a proper fix where that will hopefully not be necessary...

**Comment by damyanp** [comment]
Win2D version 0.0.15 now has the [RemoveFromVisualTree](http://microsoft.github.io/Win2D/html/M_Microsoft_Graphics_Canvas_CanvasAnimatedControl_RemoveFromVisualTree.htm) method on the various XAML controls.

These help to unhook the control enough that it can get garbage collected, without having to track event subscriptions yourself.

This isn't ideal and we're still investigating improved ways to address this, but hope that this helps make it easier to consume the Win2D controls.

**Comment by suprak** [comment]
What's the status of this?
(The link to the RemoveFromVisualTree method is dead.)

Memory leaks are a show stopper for us.
We wanted to use Win2D in WP8.1 application.

What do we have to do today to avoid ALL memory leaks in Win2D?

**Comment by damyanp** [comment]
Thanks - I've updated the link in that comment (we move the class to a different namespace, which broke the link).

For convenience, here is the updated location:
[http://microsoft.github.io/Win2D/html/M_Microsoft_Graphics_Canvas_UI_Xaml_CanvasAnimatedControl_RemoveFromVisualTree.htm](RemoveFromVisualTree).

**Comment by suprak** [comment]
So does this mean we still need to implement the cleanup methods mentioned above (unregistering event handlers, etc), or has that been fixed in the latest version on Win2D?

**Comment by shawnhar** [comment]
You do need to do some unregistering (typically from the page Unload handler) to break refcount cycles, but the RemoveFromVisualTree method provides an easier way to do this compared to tracking and unregistering all your event handlers.

There is sample code in the docs for RemoveFromVisualTree.  Also see any of the pages in the Example Gallery for real world usage  (we took care to make sure Example Gallery does this correctly and has no leaks).

**Comment by shawnhar** [comment]
Win2D now includes proper documentation on how to resolve this problem: http://microsoft.github.io/Win2D/html/RefCycles.htm

I'm closing this issue because we are not planning any further changes here in the short or medium term.  Long term, we are still looking into the possibility of OS changes that could enable a better solution in Win2D, but we have determined that

a. any such OS changes would not be until after Windows 10, so quite a way in the future
b. there is nothing more that Win2D itself can reasonably do without such OS changes
